### ISSUE TITLE
[Feature] Liveness Probe on multi-node etcd

### ISSUE THREAD

**Issue by ishan16696** [gardener/etcd-druid#280]
**Feature (What you would like to be added):**
Configure the correct livenessProbe in `multi-node etcd`.

**Motivation (Why is this needed?):**
Right now I found 2 ways to configure the livenessProbe:
1.  `/health`  endpoint
`/health` endpoint returns `false` if one of the following conditions is met ([source](https://github.com/etcd-io/etcd/blob/v3.4.14/etcdserver/api/etcdhttp/metrics.go#L106-L119)):
    - there is no etcd leader or leader-election is currently going on.
    - the latency of a QGET request exceeds 1sec

 **Disadvantage** of Method 1 (`/health` endpoint).
   -  If there is no Quorum, etcd-follower will get restart by kubelet but etcd's restart isn't a right behavior.
   -  If etcd cluster is overloaded which is bad, but restart will generate even more load

2.  `ETCDCTL_API=3 etcdctl endpoint health --endpoints=${ENDPOINTS} --command-timeout=Xs` 
`etcdctl endpoint health` command performs a GET on the "health" key([source](https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/ep_command.go#L126-L133))
    - fails if there is no etcd leader or leader-election is going on as I think GET request will fail if there is no etcd leader.

  **Disadvantage** of Method 2 (`etcdctl endpoint health`).
  -  If there is no Quorum, etcd-follower will get restart by kubelet but etcd's restart isn't a right behavior.

This issue of liveness is already reported on etcd [here](https://github.com/etcd-io/etcd/issues/13340) and it has been fixed by this [PR](https://github.com/etcd-io/etcd/pull/13399) but this feature will come with etcd v3.6 and we don't know the timeline when it will be released.

**What options we have:**
1. Go with the method 2(`etcdctl endpoint health`) and whenever etcd release v3.6, switch to that version and configure the correct liveness probe.
2. Other way is fork the etcd v3.4.13 and make the necessary changes and use the custom image.

**Comment by ishan16696** [comment]
cc @abdasgupta @aaronfern @timuthy

**Comment by timuthy** [comment]
Reading this I wonder about the option keeping the current liveness probe but setting `--consistency="s"`? This mimics the implementation of the linked PR https://github.com/etcd-io/etcd/pull/13399. WDYT @ishan16696?

**Comment by ishan16696** [comment]
> Reading this I wonder about the option keeping the current liveness probe but setting --consistency="s"

yes, this make sense to me .... I don’t see any major issue with this.
we can configure the livenessProbe like this making the consistency `serializable` instead of `linearizable`
```
- /bin/sh
      - -ec
      - ETCDCTL_API=3
      - etcdctl
      - --cert= ...
      - --key= ....
      - --cacert= ....
      - --endpoints=https://etcd-main-local:2379
      - get
      - foo
      - --consistency="s"
```

**Comment by ishan16696** [comment]
Hi @timuthy ,
Unfortunately this way of using `--consistency="s"` is not working for following scenario.
1. Start an etcd cluster with 3 members.
2. Add some dummy data.
3. Stop all the members
4. Start only one member
5. then try  `etcdctl  get foo --endpoints=http://127.0.0.1:2379 --consistency="s"`. 
Unfortunately `--consistency="s"` won't help us in this case until quorum will get satisfy.

**Comment by ishan16696** [comment]
> this last running pod is not killed by the liveness probe as well because it doesn't make sense.

I'm not saying last pod is killed due to liveness probe ... I'm saying last pod can be killed/crash due to some reason, why you are assuming that last pod can’t be killed.

**Comment by ishan16696** [comment]
The point I'm making is, suppose all pods/member of etcd-cluster crashed/go down due to some reasons, they will not be able to come up due to this livenessProbe: `etcdctl get foo  --consistency="s"` will get fail.

**Comment by ishan16696** [comment]
> (initialDelaySeconds + periodSeconds * failureThreshold)

yeah  exactly but then we have to depend on this configuration and set them accordingly.

**Comment by unmarshall** [comment]
IMO using `--consistency=s` is non-deterministic as the time it takes to start the etcd process cannot be guaranteed. The correct thing to do would be to create a `/livez` or `/healthz` endpoint which is served out of backup-restore container. If this endpoint has come up and responds with a 200 OK then etcd pod should be considered as LIVE which then prevents it from getting restarted. For readiness probe you can probably use `etcdctl` with `--consistency=s`

**Comment by timuthy** [comment]
> IMO using --consistency=s is non-deterministic as the time it takes to start the etcd process cannot be guaranteed. The correct thing to do would be to create a /livez or /healthz endpoint which is served out of backup-restore container. 

Can you sketch how an implementation of `/livez` would look like to make it deterministic?

**Comment by ishan16696** [comment]
Reopening this issue as `livenessProbe` can be improved to resolve this kind of scenario https://github.com/gardener/etcd-druid/issues/280#issuecomment-1143639829

**Comment by timuthy** [comment]
I'd be interested to know what it planned here. Do you already have something in mind?

**Comment by ishan16696** [comment]
ok, then feel free to close.

**Comment by ishan16696** [comment]
Found out the [upstream issue](https://github.com/etcd-io/etcd/issues/13524) which is also described in this comment https://github.com/gardener/etcd-druid/issues/280#issuecomment-1143639829
Mentioning this for future reference whenever we want to move to `etcd v3.6`

**Comment by ishan16696** [comment]
We had introduced the livenessProbe in multi-node etcd with this PR: https://github.com/gardener/etcd-druid/pull/396
But we found out that livenessProbe was causing more harm than good. Then, we decided to remove the livenessProbe from etcd cluster with this PR: https://github.com/gardener/etcd-druid/pull/423
So, I'm opening this issue just to track the liveness Probe for etcd cluster and whenever we will upgrade our etcd version then we can introduce the livenessProbe. Please refer to this upstream PR: https://github.com/etcd-io/etcd/pull/13525


/on-hold

**Comment by ishan16696** [comment]
This issue needs to pick after https://github.com/gardener/etcd-druid/issues/445 etcd version upgrade.

**Comment by ishan16696** [comment]
As Etcd has been upgraded to `v3.5.27` https://github.com/gardener/etcd-druid/issues/445, now livenessProbe: [/livez](https://github.com/etcd-io/etcd/blob/62d8759b7d5dbc9b3694f89d54170a55726bb485/server/etcdserver/api/etcdhttp/health.go#L277-L281) endpoint can be defined in etcd member pod as currently we don't enable or define any livenessProbe for etcd cluster members.

For more info check Health Check section in this doc: https://etcd.io/docs/v3.5/op-guide/monitoring/

**Comment by ishan16696** [comment]
I did some testing by adding a livenessProbe: [/livez](https://github.com/etcd-io/etcd/blob/62d8759b7d5dbc9b3694f89d54170a55726bb485/server/etcdserver/api/etcdhttp/health.go#L277-L281) endpoint in a etcd statefulset with following config:
```yaml
    livenessProbe:
      failureThreshold: 5
      httpGet:
        path: /livez
        port: 2379
        scheme: HTTP
      initialDelaySeconds: 50
      periodSeconds: 5
      successThreshold: 1
      timeoutSeconds: 1
```

- But I found introduction of livenessProbe can do more harm then being useful as failure of livenessProbe will simply restart the etcd container which will not helpful in many of the cases. 
- Another issue is to find the correct time for `livenessProbe.initialDelaySeconds` as sometimes etcd member can take more time in coming up due to any external (infra) issues etc. Take example of this I tried bring up the etcd cluster with `livenessProbe.initialDelaySeconds` set to `50sec` and etcd cluster member couldn't come-up as etcd container was getting restarted due to failure in livenessProbe.
```
NAME                          READY   STATUS    RESTARTS      AGE
etcd-main-0                   1/2     Running   2 (69s ago)   3m38s
etcd-main-1                   1/2     Running   2 (70s ago)   3m38s
etcd-main-2                   1/2     Running   2 (70s ago)   3m38s

k describe pod etcd-main-0
....
  Normal   Killing    9m35s (x17 over 58m)   kubelet            Container etcd failed liveness probe, will be restarted
```
- Even If I increase the `initialDelaySeconds` to some big number say `150sec` , yes etcd cluster will come up but take a scenario when etcd cluster member are busy due to high load because of this if any of member unable to respond to livenessProbe on time which can lead to failure of livenessProbe, due to that kubelet will kills the etcd container in order to restart the container.

**Conclusion**: I found livenessProbe isn't very helpful as it just kills the etcd container on failure which will not helpful because restarting a member will only worsen the situation.

**Comment by ishan16696** [comment]
I would propose to capture these findings in some doc and close this issue.

**Comment by ishan16696** [comment]
Opened a PR to update the doc with our findings of the livenessProbe configurations: https://github.com/gardener/etcd-druid/pull/1361

**Comment by ishan16696** [comment]
We have decided not to introduce the livenessProbe for etcd cluster and we have captured our decision and findings here: https://github.com/gardener/etcd-druid/blob/master/docs/concepts/etcd-cluster-components.md#livenessprobe
Nothing more to be done here, hence closing this issue, please feel free to reopen it if you feel otherwise.
/close
